### PR TITLE
feat(rollup): handle chain notifications

### LINKED ISSUE #32: feat(rollup): handle exex notifications
- after #8 
- after #31 

After initializing the pipeline, the main loop will involve taking incoming `Chain` notifications from the `Driver` and:

- handling both the committed and reverted cases
- updating the providers with the newly committed chain and blobs
- sending the `FinishedHeight` notification to the `Driver`

here is a reference implementation: 

```rust
async fn handle_exex_notification(
        &mut self,
        notification: ExExNotification,
        pipeline: &mut RollupPipeline,
    ) -> Result<()> {
        if let Some(reverted_chain) = notification.reverted_chain() {
            self.chain_provider.commit(reverted_chain.clone());
            let l1_block_info = info_from_header(&reverted_chain.tip().block);

            // handle the reverted chain...
        }

        if let Some(committed_chain) = notification.committed_chain() {
            let tip_number = committed_chain.tip().number;
            self.chain_provider.commit(committed_chain);

            if let Err(err) = self.ctx.events.send(ExExEvent::FinishedHeight(tip_number)) {
                bail!("Critical: Failed to send ExEx event: {:?}", err);
            }
        }

        Ok(())
    }
```

### PR DESCRIPTION
**Pull request by merklefruit**
closes #32

### CODE REVIEW

File: crates/rollup/src/driver.rs [hunk 1]
```diff
@@ -147,14 +168,164 @@ where
...
+        }
+
+        if let Some(committed_chain) = notif.committed_chain() {
+            let tip_number = committed_chain.tip().number;
+            // self.chain_provider.commit_chain(committed_chain);
+
+            if let Err(err) = self.ctx.send_event(ExExEvent::FinishedHeight(tip_number)) {
+                bail!("Critical: Failed to send ExEx event: {:?}", err);
+            }
+        }
```
Comment (refcell) [inline crates/rollup/src/driver.rs:208]: I think we ought to match on the `ExExNotification`. It's not marked as `non_exhaustive`, which leads me to believe new variants won't be added without a major (maybe minor) version bump.

Additionally, it looks like we're missing the case where the chain is reorged. See [`ExExNotification::ChainReorged`](https://github.com/paradigmxyz/reth/blob/51524f2534bf9b28c87be4321b6dd41378caf2f8/crates/exex/types/src/notification.rs#L14-L20). In the re-orged case, it looks like we'll want to add the new chain to the local provider, and then then _also_ walk back the old chain like logic you wrote above does for the `ChainReverted` notification.

Comment (refcell) [inline crates/rollup/src/driver.rs:208]: I think we ought to match on the `ExExNotification`. It's not marked as `non_exhaustive`, which leads me to believe new variants won't be added without a major (maybe minor) version bump.

Additionally, it looks like we're missing the case where the chain is reorged. See [`ExExNotification::ChainReorged`](https://github.com/paradigmxyz/reth/blob/51524f2534bf9b28c87be4321b6dd41378caf2f8/crates/exex/types/src/notification.rs#L14-L20). In the re-orged case, it looks like we'll want to add the new chain to the local provider, and then then _also_ walk back the old chain like logic you wrote above does for the `ChainReverted` notification.

File: crates/rollup/src/driver.rs [hunk 2]
```diff
@@ -147,14 +168,164 @@ where
         )
     }
 
-    /// Starts the Hera Execution Extension loop.
+    /// Advance the pipeline to the next L2 block.
+    async fn step(&mut self, pipeline: &mut RollupPipeline<CP, BP, L2CP>) -> Result<()> {
+        let (l2_tip, l1_origin) = self.cursor.tip();
+        let _ = pipeline.step(l2_tip).await;
+        self.cursor.advance(l1_origin, l2_tip);
```
Comment (refcell) [inline crates/rollup/src/driver.rs:175]: This is just stubbed out atm right?

File: crates/rollup/src/driver.rs [hunk 3]
```diff
@@ -147,14 +168,164 @@ where
...
+        info!("L1 chain synced to the rollup genesis block");
+
+        // Step 2: Initialize the rollup pipeline
+        let mut pipeline = self.init_pipeline();
 
-        let _pipeline = self.init_pipeline();
+        // Step 3: Start processing events
+        loop {
+            // TODO: handle pipeline step (stubbed)
+            let _ = self.step(&mut pipeline).await;
```
Comment (refcell) [inline crates/rollup/src/driver.rs:232]: Ah looks like it - nice

**Review by refcell — CHANGES_REQUESTED** [review]
Nice work!

File: crates/rollup/src/driver.rs (hunk 2 above)
Comment (merklefruit) [reply crates/rollup/src/driver.rs:175]: yep correct

File: crates/rollup/src/driver.rs (hunk 1 above)
Comment (merklefruit) [reply crates/rollup/src/driver.rs:208]: > Additionally, it looks like we're missing the case where the chain is reorged. See [ExExNotification::ChainReorged](https://github.com/paradigmxyz/reth/blob/51524f2534bf9b28c87be4321b6dd41378caf2f8/crates/exex/types/src/notification.rs#L14-L20). In the re-orged case, it looks like we'll want to add the new chain to the local provider, and then then also walk back the old chain like logic you wrote above does for the ChainReverted notification.

That logic makes total sense, but it seems like the `committed` and `reverted` notifications actually both include the `ChainReorged` notification: https://github.com/paradigmxyz/reth/blob/51524f2534bf9b28c87be4321b6dd41378caf2f8/crates/exex/types/src/notification.rs#L33, each with one half (`reverted` being the one reorged out, and `committed` being the new one)

So right now if we get a `ChainReorged` notification we'll do the following:
1. the first block: `reverted_chain()` will walk back and reset the pipeline
2. the second block: `committed_chain` will commit the new valid chain to the provider

Comment (refcell) [reply crates/rollup/src/driver.rs:208]: Oh sweet! I did not realize this, the method names are misleading. That looks good to me then!

**Review by refcell — APPROVED** [review]

**Pull request merged**